golang/btcspv/types_test.go

```diff
@@ -171,3 +171,78 @@ func (suite *TypesSuite) TestValidateSPVProof() {
+
+	digest, err := NewHash160Digest(input)
+	suite.Nil(err)
+	suite.Equal(digest, output)
+
+	badLengthInput := input[0:18]
+	_, err = NewHash160Digest(badLengthInput)
+	suite.EqualError(err, "Expected 20 bytes in a Hash160Digest, got 18")
+}
+
+func (suite *TypesSuite) TestNewHash256Digest() {
+	input := DecodeIfHex("0x1406e05881e299367766d313e26c05564ec91bf721d31726bd6e46e60689539a")
+	output := Hash256Digest{0x14, 0x06, 0xe0, 0x58, 0x81, 0xe2, 0x99, 0x36, 0x77, 0x66, 0xd3, 0x13, 0xe2, 0x6c, 0x05, 0x56, 0x4e, 0xc9, 0x1b, 0xf7, 0x21, 0xd3, 0x17, 0x26, 0xbd, 0x6e, 0x46, 0xe6, 0x06, 0x89, 0x53, 0x9a}
+
+	digest, err := NewHash256Digest(input)
+	suite.Nil(err)
+	suite.Equal(digest, output)
+
+	input = DecodeIfHex("0x4f8b42c22dd3729b519ba6f68d2da7cc5b2d606d05daed5ad5128cc03e6c6358")
+	output = Hash256Digest{0x4f, 0x8b, 0x42, 0xc2, 0x2d, 0xd3, 0x72, 0x9b, 0x51, 0x9b, 0xa6, 0xf6, 0x8d, 0x2d, 0xa7, 0xcc, 0x5b, 0x2d, 0x60, 0x6d, 0x05, 0xda, 0xed, 0x5a, 0xd5, 0x12, 0x8c, 0xc0, 0x3e, 0x6c, 0x63, 0x58}
+
+	digest, err = NewHash256Digest(input)
+	suite.Nil(err)
+	suite.Equal(digest, output)
+
+	badLengthInput := input[0:30]
+	_, err = NewHash256Digest(badLengthInput)
+	suite.EqualError(err, "Expected 32 bytes in a Hash256Digest, got 30")
+}
+
+func (suite *TypesSuite) TestNewRawHeader() {
+	input := DecodeIfHex("0x7bb2b8f32b9ebf13af2b0a2f9dc03797c7b77ccddcac75d1216389abfa7ab3750000000000ffffffffaa15ec17524f1f7bd47ab7caa4c6652cb95eec4c58902984f9b4bcfee444567d0000000000ffff")
+	output := RawHeader{0x7b, 0xb2, 0xb8, 0xf3, 0x2b, 0x9e, 0xbf, 0x13, 0xaf, 0x2b, 0x0a, 0x2f, 0x9d, 0xc0, 0x37, 0x97, 0xc7, 0xb7, 0x7c, 0xcd, 0xdc, 0xac, 0x75, 0xd1, 0x21, 0x63, 0x89, 0xab, 0xfa, 0x7a, 0xb3, 0x75, 0x00, 0x00, 0x00, 0x00, 0x00, 0xff, 0xff, 0xff, 0xff, 0xaa, 0x15, 0xec, 0x17, 0x52, 0x4f, 0x1f, 0x7b, 0xd4, 0x7a, 0xb7, 0xca, 0xa4, 0xc6, 0x65, 0x2c, 0xb9, 0x5e, 0xec, 0x4c, 0x58, 0x90, 0x29, 0x84, 0xf9, 0xb4, 0xbc, 0xfe, 0xe4, 0x44, 0x56, 0x7d, 0x00, 0x00, 0x00, 0x00, 0x00, 0xff, 0xff}
+
+	header, err := NewRawHeader(input)
+	suite.Nil(err)
+	suite.Equal(header, output)
+
+	badLengthInput := input[0:70]
+	_, err = NewRawHeader(badLengthInput)
+	suite.EqualError(err, "Expected 80 bytes in a RawHeader got 70")
+}
+
+func (suite *TypesSuite) TestHeaderFromRaw() {
+	validHeader := suite.ValidProof.ConfirmingHeader
+	// PrevHash is stored in JSON as BE, we need to reverse it before comparing
+	reversed, _ := NewHash256Digest(ReverseEndianness(validHeader.PrevHash[:]))
+	validHeader.PrevHash = reversed
+	var height uint32 = 592920
+
+	rawHeader := HeaderFromRaw(validHeader.Raw, height)
+	suite.Equal(rawHeader, validHeader)
+}
+
+func (suite *TypesSuite) TestHeaderFromHex() {
+	hex := "0x0000c020c238b601308b7297346ab2ed59942d7d7ecea8d23a1001000000000000000000b61ac92842abc82aa93644b190fc18ad46c6738337e78bc0c69ab21c5d5ee2ddd6376d5d3e211a17d8706a84"
+	var height uint32 = 592920
```

Comment:
can we get this from the test suite instead of hardcoding?

traverse the valid header array and make sure each one can be generated from its own `raw` and `height` fields?

Reply:
Same goes for all hard-coded constants

Reply:
Right now we only have one valid header, but I can make it so if we add more, it will test all of them.

Reply:
Fixed this, now getting these values from the test suite